packages/tanstack-react-query/src/internals/createOptionsProxy.ts

```diff
@@ -46,9 +84,33 @@ export interface DecorateQueryProcedure<TDef extends ResolverDef> {
    * @see https://tanstack.com/query/latest/docs/framework/react/reference/infiniteQueryOptions#infinitequeryoptions
    */
   infiniteQueryOptions: TRPCInfiniteQueryOptions<TDef>;
+
+  /**
+   * Calculate the Tanstack Query Key for a Query Procedure
+   *
+   * @see https://tanstack.com/query/latest/docs/framework/react/guides/query-keys
+   */
+  queryKey: (input?: TDef['input']) => TRPCQueryKey;
+
+  /**
+   * Calculate a Tanstack Query Filter for a Query Procedure
+   *
+   * @see https://tanstack.com/query/latest/docs/framework/react/guides/filters
+   */
+  queryFilter: (input?: TDef['input']) => QueryFilters;
 }
 
 export interface DecorateMutationProcedure<TDef extends ResolverDef> {
+  /**
+   * @internal prefer using InferInput to access input type
+   */
+  '~input': TDef['input'];
+
+  /**
+   * @internal prefer using InferOutput to access input type
+   */
+  '~output': TDef['output'];
```

Comment:
nit: we usually use `$` to signify types

Reply:
could maybe be

```ts
$types: {
  input: x
  output: y
}
```
to align with e.g. our root obj

Reply:
i think the promise was that ~ gets sorted last in auto-complete, but I can't repro that 🤔 

![CleanShot 2024-11-17 at 22 20 45](https://github.com/user-attachments/assets/6455a689-0ea6-4950-934c-8115adefa6b0)

Reply:
Pushed this as the probable outcome but happy to discuss further

<img width="582" alt="image" src="https://github.com/user-attachments/assets/3c3925ac-7221-408b-97cd-e6eaaf4123b7">

Reply:
@juliusmarminge I think the sorting trick doesn't work in object autocomplete but it does for autocompleting chaining, see:

<img width="493" alt="image" src="https://github.com/user-attachments/assets/303594db-029c-4507-bffe-72b36c776fb8">

Reply:
ah okay